### PR TITLE
feat(eslint-plugin): add rule `non-nullable-type-assertion-style`

### LINKED ISSUE #202: Rule proposal: Prefer non-null assertion operator over type assertion
I have the following code:

```ts
function foo(count: number) {}

// …

foo(count: lookup);
```

Where `lookup` is `number | undefined`.

In this code path, I know for sure that `lookup` will be a number, so I have done the following:

```ts
foo(count: lookup as number);
```

But a better solution would have been to use the [non-null assertion operator](https://stackoverflow.com/questions/42273853/in-typescript-what-is-the-exclamation-mark-bang-operator-when-dereferenci):

```ts
foo(count: lookup!);
```

I would like a rule that enforces this.

---

It should be noted in the rule docs that this conflicts with the `no-non-null-assertion` rule.

Alternatively, the `no-non-null-assertion` could be renamed to `non-null-assertion` and handle both ways. That might be even better.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Fixes #202.

### CODE REVIEW

**Comment by JoshuaKGoldberg** [comment]
Err, `$ git diff --quiet --exit-code` is failing in CI and I don't know why. 😕

File: packages/eslint-plugin/typings/typescript.d.ts [hunk 1]
```diff
@@ -25,6 +25,15 @@ declare module 'typescript' {
...
 
   interface Type {
+    /**
+     * Private, unique identifier for the type as cached by the TypeScript program.
+     *
+     * @remarks
+     * This should rarely be used, and then only to check for reference equality
+     * between types (since we don't have an available isAssignableTo API).
+     */
+    id?: string;
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/typings/typescript.d.ts:35]: I'm not too pleased with relying on this `id` to exist... but short of a type relationship API in TypeScript (https://github.com/microsoft/TypeScript/issues/9879), I don't see a better way...?

Comment (bradzacher) [reply packages/eslint-plugin/typings/typescript.d.ts:35]: unless I'm mistaken - I think that you should be able to use `===` to compare.

File: packages/eslint-plugin/src/rules/non-nullable-type-assertion-style.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      const type = checker.getTypeAtLocation(
+        parserServices.esTreeNodeToTSNodeMap.get(node),
+      );
+
+      return tsutils.isTypeFlagSet(
+        type,
+        ts.TypeFlags.Any | ts.TypeFlags.Unknown,
+      )
+        ? undefined
+        : tsutils.unionTypeParts(type);
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/non-nullable-type-assertion-style.ts:43]: nitpick - IMO it's clearer without a ternary

```suggestion
      if (
        tsutils.isTypeFlagSet(type, ts.TypeFlags.Any | ts.TypeFlags.Unknown)
      ) {
        return undefined;
      }
      return tsutils.unionTypeParts(type);
```

**Review by bradzacher — CHANGES_REQUESTED** [review]
Code itself LGTM.

----

I wonder if this rule could also check for this case:
```ts
type T = string | null;
declare const x: T;

const y = x as NonNullable<T>;
// should be
const y = x!;
```

File: packages/eslint-plugin/typings/typescript.d.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/typings/typescript.d.ts:35]: Ha! I could have sworn I'd tried that and seen evidence to the contrary. No longer. Excellent. Thank you 😄.

**Comment by JoshuaKGoldberg** [comment]
Adding `NonNullable` SGTM - I'll be mostly out of availability the next couple of weeks, but hopefully will either find time to sneak it in with then or should be able to get to later. 👍

**Comment by JoshuaKGoldberg** [comment]
Amusingly, even though I took a month and a half to get back to this, it's already supported: the logic just checks for whether the types are the same but for `null` and/or `undefined` added. So `NonNullable` and any of equivalents are supported already. Added test cases!

**Review by bradzacher — APPROVED** [review]
LGTM - thanks for your work!

**Pull request merged**